dpnp/dpnp_utils/dpnp_utils_linearalgebra.py

```diff
@@ -925,17 +972,125 @@ def dpnp_matmul(
-            return dpnp.asarray(result, order=order)
-        # dpnp.ascontiguousarray changes 0-D array to 1-D array
-        if result.ndim == 0:
-            return result
-        return dpnp.ascontiguousarray(result)
+        return dpnp.asarray(result, order=order)
 
     result = dpnp.get_result_array(result, out, casting=casting)
     if axes is not None and out is result:
         # out and out_orig contain the same data but they have different shape
         return out_orig
     return result
+
+
+def dpnp_vecdot(
+    x1,
+    x2,
+    /,
+    out=None,
+    *,
+    casting="same_kind",
+    order="K",
+    dtype=None,
+    axes=None,
+    axis=None,
+):
+    """Vector dot product of two arrays."""
+
+    dpnp.check_supported_arrays_type(x1, x2)
+    res_usm_type, exec_q = get_usm_allocations([x1, x2])
+    _validate_out_array(out, exec_q)
+
+    if order in "aAkK":
+        if x1.flags.fnc and x2.flags.fnc:
+            order = "F"
+        else:
+            order = "C"
+
+    x1_ndim = x1.ndim
+    x2_ndim = x2.ndim
+    if axes is None and axis is None:
+        # default behavior with axis=-1
+        pass
+    elif axes is not None:
+        if axis is not None:
+            raise TypeError("cannot specify both `axis` and `axes`.")
+
+        axes = _validate_axes(x1, x2, axes, "vecdot")
+
+        axes_x1, axes_x2, axes_res = axes
+        axes_x1 = normalize_axis_tuple(axes_x1, x1_ndim, "axis")
+        axes_x2 = normalize_axis_tuple(axes_x2, x2_ndim, "axis")
+
+        # Move the axes that are going to be used in dot product,
+        # to the end of "x1" and "x2"
+        x1 = dpnp.moveaxis(x1, axes_x1, -1) if x1_ndim != 1 else x1
+        x2 = dpnp.moveaxis(x2, axes_x2, -1) if x2_ndim != 1 else x2
+    else:
+        x1 = dpnp.moveaxis(x1, axis, -1) if x1_ndim != 1 else x1
+        x2 = dpnp.moveaxis(x2, axis, -1) if x2_ndim != 1 else x2
+
+    # When inputs are 1-D arrays, the result is a 0-D array. For this case,
+    # NumPy allows out keyword to have any shape and the result is broadcast to it
+    NumPy_special_behavior = (
+        out is not None and x1_ndim == 1 and x2_ndim == 1 and out.shape != ()
+    )
+
+    x1, x2, result_shape = _get_result_shape(
+        x1, x2, out, "vecdot", NumPy_special_behavior
+    )
+
+    # Determine the appropriate data types
+    _, res_dtype = _compute_res_dtype(
+        x1, x2, dtype=dtype, casting=casting, sycl_queue=exec_q
+    )
+
+    _, x1_is_1D, _ = _define_dim_flags(x1, axis=-1)
+    _, x2_is_1D, _ = _define_dim_flags(x2, axis=-1)
+
+    if x1.size == 0 or x2.size == 0:
+        order = "C" if order in "kK" else order
+        result = _create_result_array(
+            x1,
+            x2,
+            out,
+            shape=result_shape,
+            dtype=res_dtype,
+            usm_type=res_usm_type,
+            sycl_queue=exec_q,
+            order=order,
+        )
+        if numpy.prod(result_shape) == 0:
+            return result
+        result.fill(0)
```

Comment:
What if `out` keyword is passed? The behavior might be incorrect. like below:
```python
ia = dpnp.ones((2, 0), dtype='f8')
ib = dpnp.ones((2, 0), dtype='f8')
ic = dpnp.ones(2, dtype='i8')

dpnp.vecdot(ia, ib, out=ic)
# Out: array([0., 0.])
ic
# Out: array([1, 1])

c = numpy.ones(2)
numpy.vecdot(ia.asnumpy(), ib.asnumpy(), out=c)
# Out: array([0., 0.])

c
# Out: array([0., 0.])

``

Reply:
Updated the code to fix this issue. Note the `dpnp` and NumPy` examples provided here are not exactly the same and `dpnp` example (an a similar `NumPy`) will raise an error because of `casting` error.